src/forder.c

```diff
@@ -801,6 +800,145 @@ SEXP forder(SEXP DT, SEXP by, SEXP retGrpArg, SEXP sortGroupsArg, SEXP ascArg, S
+    }
+  }
+  UNPROTECT(1);
+  return true;
+}
+
+// paste0("__", cols, collapse="")
+SEXP idxName(SEXP cols) {
+  if (!isString(cols))
+    error("internal error: 'cols' must be a character"); // # nocov
+  SEXP char_underscore2 = PROTECT(ScalarString(mkChar("__")));
+  SEXP char_empty = PROTECT(ScalarString(mkChar("")));
+  SEXP sym_paste0 = install("paste0");
+  SEXP call_paste0 = PROTECT(lang4(sym_paste0, char_underscore2, cols, char_empty));
+  SET_TAG(CDDDR(call_paste0), install("collapse"));
+  SEXP ans = PROTECT(eval(call_paste0, R_GlobalEnv));
+  UNPROTECT(4);
+  return ans;
+}
+
+// attr(attr(x, "index"), idxName(names(x)[cols]))
+SEXP getIndex(SEXP x, SEXP cols) {
+  if (!isInteger(cols))
+    error("internal error: 'cols' must be an integer"); // # nocov
+  SEXP index = getAttrib(x, sym_index);
+  if (isNull(index))
+    return index;
+  SEXP idx_names = PROTECT(subsetVector(getAttrib(x, R_NamesSymbol), cols));
+  SEXP name_idx = PROTECT(idxName(idx_names));
+  SEXP sym_idx = install(CHAR(STRING_ELT(name_idx, 0)));
+  SEXP idx = getAttrib(index, sym_idx);
+  UNPROTECT(2);
+  return idx;
+}
+
+// lazy forder, re-use existing key or index if possible, otherwise call forderDo
+SEXP forder(SEXP DT, SEXP by, SEXP retGrpArg, SEXP sortGroupsArg, SEXP ascArg, SEXP naArg, SEXP lazyArg) {
+  const bool verbose = GetVerbose();
+  int protecti = 0;
+  double tic=0.0;
+  if (verbose)
+    tic = omp_get_wtime();
+  if (isNull(DT))
+    error("DT is NULL");
+  if (!IS_TRUE_OR_FALSE(retGrpArg))
+    error("retGrp must be TRUE or FALSE");
+  if (!IS_TRUE_OR_FALSE(sortGroupsArg))
+    error("sort must be TRUE or FALSE");
+  if (!isLogical(naArg) || LENGTH(naArg) != 1)
+    error("na.last must be logical TRUE, FALSE or NA of length 1");
+  if (!isInteger(ascArg))
+    error("order must be integer"); // # nocov # coerced to int in R
+  if (!isLogical(lazyArg) || LENGTH(lazyArg) != 1)
+    error("lazy must be logical TRUE, FALSE or NA of length 1");
+
+  if (!length(DT))
+    return allocVector(INTSXP, 0);
+
+  int opt = -1; // -1=unknown, 0=none, 1=keyOpt, 2=idxOpt
+  if (LOGICAL(lazyArg)[0]==NA_LOGICAL) {
+    if (isNewList(DT) &&
+        !LOGICAL(retGrpArg)[0] &&
+        LOGICAL(sortGroupsArg)[0] &&
+        !LOGICAL(naArg)[0] &&
+        all1(ascArg)) {
+      opt = -1;
+    } else {
+      opt = 0;
+    }
+  } else if (LOGICAL(lazyArg)[0]) {
+    if (!isNewList(DT))
+      error("internal error: lazy set to TRUE but DT is not a list"); // # nocov
+    if (LOGICAL(retGrpArg)[0])
+      error("internal error: lazy set to TRUE but retGrp is TRUE"); // # nocov # proposal to change that: #4346
+    if (!LOGICAL(sortGroupsArg)[0])
+      error("internal error: lazy set to TRUE but sort is FALSE"); // # nocov
+    if (LOGICAL(naArg)[0]!=FALSE)
+      error("internal error: lazy set to TRUE but na.last is not FALSE"); // # nocov
+    if (!all1(ascArg))
+      error("internal error: lazy set to TRUE but order is not all 1"); // # nocov
+    opt = -1;
```

Comment:
wherever we call now R's `forderv` with an intent to calculate order (not `na.last` etc), we could pass `lazy=TRUE`